fedbiomed/node/node_state_manager.py

```diff
@@ -0,0 +1,262 @@
+# This file is originally part of Fed-BioMed
+# SPDX-License-Identifier: Apache-2.0
+
+import os
+from typing import Dict, Optional, Union, Any
+import uuid
+
+from tinydb import TinyDB, Query
+from tinydb.table import Table
+
+from fedbiomed.common.utils import raise_for_version_compatibility
+from fedbiomed.common.constants import (_BaseEnum, ErrorNumbers, NODE_STATE_PREFIX,
+                                        __node_state_version__)
+from fedbiomed.common.exceptions import FedbiomedNodeStateManagerError
+from fedbiomed.common.logger import logger
+
+from fedbiomed.node.environ import environ
+
+NODE_STATE_TABLE_NAME = "Node_states"
+
+
+class NodeStateFileName(_BaseEnum):
+    """
+    Collection of File names that composes a `Node` state.
+    File names should contains 2 `%s`: one for round number, the second for state_id
+    Example: VALUE = "some_values_%s_%s"
+    """
+    OPTIMIZER: str = "optim_state_%s_%s"
+
+
+class NodeStateManager:
+    """Node state saving facility: Handles saving and loading Node states from previous `Rounds`,
+    given a `state_id`.
+    Interfaces with database use to save and load Node State entries.
+
+    Database Table that handles Node states is built with the following entries:
+        - state_id
+        - job_id
+        - version_node_id: state id version, that checks if states can be used regarding FedBioMed versions.
+        - state_entries: content of each entry that cmposes a Node State
+    """
+
+    def __init__(self, db_path: str):
+        """Constructor of the class.
+
+        Args:
+            db_path: path to the node state database
+
+        Raises:
+            FedbiomedNodeStateManagerError: failed to access the database
+
+    """
+        # NOTA: constructor has been designed wrt other object handling DataBase
+        self._query: Query = Query()
+        # node state base directory, where all node state related files are saved
+        self._node_state_base_dir: Optional[str] = None
+        self._state_id: Optional[str] = None
+        self._previous_state_id: Optional[str] = None
+        try:
+            self._db: Table = TinyDB(db_path).table(name=NODE_STATE_TABLE_NAME, cache_size=0)
```

Comment:
Small comment on `tinydb.Table`: All other DB implementations uses DBTable that is an interface of `Table`.  There two options to make tinydb to use custom table class 

```
self._db = TinyDB(db_path)
self._table = DBTable(name=NODE_STATE_TABLE_NAME, cache_size=0)
```

or 

```
connection = TinyDB(db_path)
connection.table_class = DBTable 
self._db = connection.table(name=NODE_STATE_TABLE_NAME, cache_size=0)
```